hikari/impl/rest.py

```diff
@@ -2095,9 +2095,20 @@ async def fetch_ban(
         self,
         guild: snowflakes.SnowflakeishOr[guilds.PartialGuild],
         user: snowflakes.SnowflakeishOr[users.PartialUser],
-    ) -> guilds.GuildMemberBan:
+    ) -> typing.Optional[guilds.GuildMemberBan]:
         route = routes.GET_GUILD_BAN.compile(guild=guild, user=user)
-        raw_response = await self._request(route)
+        try:
+            raw_response = await self._request(route)
+        except errors.NotFoundError as ex:
+            # Discord raises an error when the ban is not found.
+            # We simplify this for the user and only raise the
+            # NotFoundError if it is something else not being found.
+            json_body = typing.cast(data_binding.JSONObject, data_binding.load_json(ex.raw_body))
```

Comment:
Should we be doing this for other endpoints too, then?

I wonder if this is a little inconsistent with other endpoints, is all.

I wonder whether it would be worth adding an optional `type` field to `NotFoundError` instead so we can flag to the user what is missing.

Maybe something like uh...

```py
class NotFoundError(...):
    type: typing.Optional[typing.Type[typing.Any]]
```
and then
```py
if json_body["code"] == xxxx:
    raise NotFoundError(guilds.GuildMemberBan)
```
I'm not sure.

On the other hand, your way makes sense in my head. I'm just not sure if we should raise in some cases but not others. I'll have a ponder.

Reply:
Actually, I think we should approach this differently.

We should get Discord to verify the basic format for exceptions, so we know it is behaviour we can use without risking it breaking. I'll assume we can only expect a code in it and should ignore anything else.

We should then add the JSON error code enum to errors.py. The code should then be an optional field on any HTTP exception. This will allow users to instead check the exact nature of the error and this will then be applicable to all endpoints.

I think that is the best way forwards with this. It would also possibly allow us to tidy up the str/repr for http errors.